h/traversal/group_membership.py

```diff
@@ -0,0 +1,60 @@
+from dataclasses import dataclass
+
+from h.exceptions import InvalidUserId
+from h.models import Group, GroupMembership, User
+
+
+def group_membership_api_factory(request):
+    """Route factory for the "api.group_member" route."""
+    return GroupMembershipRootNode(request)
+
+
+class GroupMembershipRootNode:
+    def __init__(self, request):
+        self.request = request
+        self.group_service = request.find_service(name="group")
+
+    def __getitem__(self, pubid):
+        group = self.group_service.fetch(pubid)
+
+        if not group:
+            raise KeyError(pubid)
+
+        return GroupMembershipGroupNode(self.request, group)
+
+
+class GroupMembershipGroupNode:
+    def __init__(self, request, group):
+        self.request = request
+        self.group = group
+        self.user_service = request.find_service(name="user")
+        self.group_members_service = request.find_service(name="group_members")
+
+    def __getitem__(self, userid):
+        if userid == "me":
+            if self.request.authenticated_userid:
+                user = self.user_service.fetch(self.request.authenticated_userid)
+            else:
+                raise KeyError(userid)
+        else:
+            try:
+                user = self.user_service.fetch(userid)
+            except InvalidUserId as exc:
+                raise KeyError(userid) from exc
+
+        if not user:
+            raise KeyError(userid)
+
+        membership = self.group_members_service.get(self.group, user)
+
+        if self.request.method != "POST" and not membership:
+            raise KeyError((self.group, user))
+
+        return GroupMembershipContext(self.group, user, membership)
+
+
+@dataclass
+class GroupMembershipContext:
+    group: Group
+    user: User
+    membership: GroupMembership | None
```

Comment:
Ok, so this uses Pyramid's "traversal" feature. You can read about traversal in several pages of the Pyramid docs:

1. [Hello Traversal World](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hellotraversal.html)
2. [Much Ado About Traversal](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/muchadoabouttraversal.html)
3. [Traversal](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/traversal.html)
4. [Combining Traversal and URL Dispatch](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hybrid.html) (which is the particular way that h uses traversal)
5. [ZODB + Traversal Wiki Tutorial](https://docs.pylonsproject.org/projects/pyramid/en/latest/tutorials/wiki/index.html)

I realise that this is very "Pyramidy" but it has some big advantages:

1. If the group, user or membership doesn't exist then the traversal or "resource lookup" code in this file will raise `KeyError` which will cause Pyramid to 404 the request without ever calling the security policy or view.
2. h can return 404s for resources that don't exist, and 401s for resources that the request doesn't have permission to act on.
3. If the security policy is called it can safely assume that the group, user and membership were found and can be accessed on the `context`.
4. If the view is called  it can safely assume that the group, user and membership were found and can be accessed on the `context` and the security policy permission check passed.

If you don't use traversal you get a problem. If someone makes an API request with an unknown group ID or unknown user ID in the URL, or with a group ID and user ID for which there is no membership, then Pyramid will end up calling the security policy which must return either `Allowed` or `Denied`, neither of which is a good choice:

1. It doesn't seem quite right to me for the security policy to return `Denied` in this situation. It's not that the request doesn't have permission to edit the membership, the problem is that the membership doesn't exist. Returning `Denied` would cause Pyramid to send a 401 response, when a 404 would be more appropriate.

2. The security policy could return `Allowed` and then the view could raise `HTTPNotFound`. This would return the correct response (404) but it seems wrong to me for the security policy to say that the request is permitted to edit a membership that doesn't exist and I imagine there could even be race conditions resulting in security issues (if a membership with the matching IDs gets created by a concurrent request and it happens to be a membership that the API request should *not* have been permitted to edit).

So, here's how this works. The route config looks like this:

```python
# routes.py

config.add_route(
    "api.group_member",
    "/api/groups/{pubid}/members/{userid}",
    factory="h.traversal.group_membership_api_factory",
    traverse="/{pubid}/{userid}",
)
```

The `factory="h.traversal.group_membership_api_factory"` means that when a request matches this route Pyramid will call `group_membership_api_factory()` which will return a `GroupMembershipRootNode` instance.

The `traverse="/{pubid}/{userid}"` means that Pyramid will call `GroupMembershipRootNode.__getitem__(pubid)` (passing the `pubid` from the request URL path) which will return a `GroupMembershipGroupNode` instance on which Pyramid will call `GroupMembershipGroupNode.__getitem__(userid)` which will return a `GroupMembershipContext` which Pyramid will then use as the `context` argument passed to the security policy, view, etc.

If either `GroupMembershipRootNode.__getitem__(pubid)` or `GroupMembershipGroupNode.__getitem__(userid)` raises `KeyError` then Pyramid will 404 the request without calling the security policy or view.

That's all there is to it. I realise it's a little weird having to define a tree of classes like this. But it's simple enough and it's a good way to separate looking up the group, user and membership (and handling when they don't exist) from making security permissions decisions and from business logic (the view function). It's a good separation of concerns.

Also as I said AFAIK it's the only way to avoid Pyramid calling the security policy when the group, user and/or membership doesn't exist and there's no sensible response for the security policy to give.